c-dependencies/js-compute-runtime/builtins/json-web-key.cpp

```diff
@@ -0,0 +1,353 @@
+    }
+    if (has_ext) {
+      JS::RootedValue ext_val(cx);
+      if (!JS_GetProperty(cx, object, "ext", &ext_val)) {
+        return nullptr;
+      }
+      ext = JS::ToBoolean(ext_val);
+    }
+  }
+  //   sequence<DOMString> key_ops;
+  std::vector<std::string> key_ops;
+  {
+    bool has_key_ops;
+    if (!JS_HasProperty(cx, object, "key_ops", &has_key_ops)) {
+      return nullptr;
+    }
+    if (has_key_ops) {
+      JS::RootedValue key_ops_val(cx);
+      if (!JS_GetProperty(cx, object, "key_ops", &key_ops_val)) {
+        return nullptr;
+      }
+      bool key_ops_is_array;
+      if (!JS::IsArrayObject(cx, key_ops_val, &key_ops_is_array)) {
+        return nullptr;
+      }
+      if (!key_ops_is_array) {
+        // TODO: Check if key_ops_val is iterable via Symbol.iterator and if so, convert to a JS
+        // Array
+        JS_ReportErrorASCII(cx, "Failed to read the 'key_ops' property from 'JsonWebKey': The "
+                                "provided value cannot be converted to a sequence");
+        return nullptr;
+      }
+      uint32_t length;
+      JS::RootedObject key_ops_array(cx, &key_ops_val.toObject());
+      if (!JS::GetArrayLength(cx, key_ops_array, &length)) {
+        return nullptr;
+      }
+      if (length != 0) {
+        // Initialize to the provided array's length.
+        // TODO: move to mozilla::Vector so we can catch if this reserve exceeds our memory limits
+        key_ops.reserve(length);
+        JS::RootedValue op_val(cx);
+        for (uint32_t i = 0; i < length; ++i) {
+          // We should check for and handle interrupts so we can allow GC to happen whilst we are
+          // iterating.
+          // TODO: Go through entire codebase and add JS_CheckForInterrupt into code-paths which
+          // iterate or are recursive such as the structuredClone function if
+          // (!JS_CheckForInterrupt(cx)) {
+          //   return nullptr;
+          // }
+
+          if (!JS_GetElement(cx, key_ops_array, i, &op_val)) {
+            return nullptr;
+          }
+
+          size_t op_length;
+          auto op_chars = encode(cx, op_val, &op_length);
+          if (!op_chars) {
+            return nullptr;
+          }
+
+          std::string op(op_chars.get(), op_length);
+
+          if (op != "encrypt" && op != "decrypt" && op != "sign" && op != "verify" &&
+              op != "deriveKey" && op != "deriveBits" && op != "wrapKey" && op != "unwrapKey") {
+            JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                                      JSMSG_SUBTLE_CRYPTO_INVALID_KEY_USAGES_VALUE);
+            return nullptr;
+          }
+
+          // No duplicates allowed
+          if (std::find(key_ops.begin(), key_ops.end(), op) != key_ops.end()) {
+            JS_ReportErrorASCII(
+                cx, "The 'key_ops' member of the JWK dictionary contains duplicate usages");
+            return nullptr;
+          }
+
+          key_ops.push_back(op);
```

Comment:
Using `push_back` here will copy the string, but `emplace_back` will give you the opportunity to invoke the move constructor for the target cell in the vector.
```suggestion
          key_ops.emplace_back(std::move(op));
```

Reply:
I partially take this back: there's no need to use `emplace_back` here, `push_back(std::move(op))` will get the same effect :+1: